### ISSUE TITLE
use get_ncol() for reading CompReport.sso

### ISSUE THREAD

**Issue by iantaylor-NOAA** [r4ss/r4ss#642]
@kellijohnson-NOAA and @k-doering-NOAA, can you help me solve a puzzle?

@cminte reported an issue with the hard-wired value for the number of columns in CompReport.sso here:
https://github.com/r4ss/r4ss/blob/17a16966e786820081685b05164f70656457b65e/R/SS_output.R#L836-L839
(where increasing the value above 300 solved the problem for her).

The history of that the hardwired value is that I first tried using the new `get_ncol()` function for CompReport in this commit from Sept 11, 2020, de18b074cb010bca42c0e6d0f476f6af4d278395, and then dialed it back and replaced by what I thought was an adequately large hardwired value in this commit https://github.com/r4ss/r4ss/commit/ee703b0124866f41c0543fd3de9c5d9e629ccc26. The error that led to the dialing back is described in the email pasted below. However, I can't replicate the error just now, and restoring the use of `get_ncol()` in the "get_ncol" branch just now passed all the tests: https://github.com/r4ss/r4ss/actions/runs/2098884621.

Are you aware of any changes to `read.table()` in more recent versions of R, or do you have ideas for how to replicate the error from 2020? And if not, should I just merge `get_ncol` and hope that the problem has gone away?

> On Fri, Sep 11, 2020 at 2:44 PM Kathryn Doering - NOAA Affiliate <kathryn.doering@noaa.gov> wrote:
> Hi Ian,
> 
>  I checked out th ss3sim failing tests and they are all failing for the same reason, so hopefully the fix is straightforward!
> 
> This is the problem line: https://github.com/r4ss/r4ss/commit/de18b074cb010bca42c0e6d0f476f6af4d278395#diff-2bcf327a64b23bdd518194b1c538e791R828
> 
> due to "more columns than column names".
> 
> I was unsure how to modify the get_ncol function to for a fix, but here is how to reproduce the error:
> Use the CompReport.sso (attached) with get_ncol("CompReport.sso", skip = 3, nrows = 25) using the development branch of r4ss. You should get the error
>   Error in utils::read.table(file, col.names = 1:initial, fill = TRUE, quote = "",  : 
>   more columns than column names 
> 
> which I think is b/c there are more than 100 columns (initial is hard coded as 100 to start).
> 
> Do you see an easy fix for this problem?  If not, I can spend more time trying to figure it out.
> 
> Thanks,
> Kathryn

**Comment by yukio-takeuchi** [comment]
Sorry to my intervention, but I think,  r4ss:::get_ncol still does not work properly.
For example, I got a same error as reported previously when I applied get_ncol to a Comp/report.sso as
```
➜ Rscript -e "r4ss:::get_ncol('CompReport.sso', skip = 3, nrows = 25)"
Error in utils::read.table(file, col.names = 1:initial, fill = TRUE, quote = "",  :
  more columns than column names
Calls: <Anonymous> -> <Anonymous>
Execution halted
```
I reviewed the source code of read.table. This error was caused if maximum number of columns of 25 rows  are more than length(col.names). If col.names is not given, read.table estimates number of columns by max(number of columns of skip+1:nrows lines of the file).
Simple solution would be to apply read.table without col.names.
e.g.
```
❯ Rscript -e "max(apply(read.table('CompReport.sso', skip = 3, nrows = 25,fill=T), 1, length))"
[1] 105
```

**Comment by iantaylor-NOAA** [comment]
@yukio-takeuchi, thanks for digging into this. Your input is ALWAYS welcome.
I pushed the change you suggested to the `get_ncols` branch to confirm that it passes all the tests (https://github.com/r4ss/r4ss/actions?query=branch%3Aget_ncol).

**Comment by kellijohnson-NOAA** [comment]
YES! 🥇 thank you @yukio-takeuchi your fix allowed me time to work on other things in {r4ss} and {ss3sim} today!

**Comment by yukio-takeuchi** [comment]
I think recursive application of read.table is unnecessary. Without col.names rread.table applies "scan" for each single line in lines skip +1 to skip + nrows, then number of columns are determined as maximum of number of columns of  lines read. Therefore single application of read.table should be enough to determine number of columns

**Comment by iantaylor-NOAA** [comment]
@yukio-takeuchi, if I understand correctly, you're suggesting that we can just use `read.table()` directly within `SS_output()` rather than wrap it up in the `get_ncol()` function. Looking at the documentation for `read.table()`, it looks like the scan for how many columns with `fill = TRUE` only applies to the first 5 rows (see https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/read.table#:~:text=The%20number%20of%20data%20columns). But if you're right that it sill scan all the way to nrows, then that should work, especially in this case where we're just trying to parse the top part of CompReport.sso.

I also see the `read.table()` [example](https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/read.table#:~:text=NOT%20RUN) proposes a solution using `count.fields()` which we could try out.

I'm on leave most of next week so it may take a while for me to contribute more on this, but there's no big rush since the temporary work around worked.

**Comment by yukio-takeuchi** [comment]
@iantaylor-NOAA 

Sorry I misunderstood the structure of get_ncol function.
read.table need to be applied iteratively. But in while block col.names=1:initial as one of argument of read.table. Otherwise read.table determine number of columns based on number of columns of first 5 lines 
 However initial value of initial (100) is the cause of error.
As I mentioned previously if number of columns determined from the first 5 lines are more than length(col.names), read.table stop with error message.
Therefore initial needs to be initialized e.g.
```
initial <- utils::read.table(file,
            fill = TRUE, quote = "", colClasses = "character", 
            nrows = nrows, skip = skip, comment.char = "")
    initial <- dim(initial)[2]
```
This allows us to determine large enogh initial value.
Revised get_ncol to be something like
```
get_ncol<-
function (file, skip = 0, nrows = -1) 
{
    numcol <- list("yes")
    initial <- utils::read.table(file,
            fill = TRUE, quote = "", colClasses = "character", 
            nrows = nrows, skip = skip, comment.char = "")
    initial <- dim(initial)[2]
    while (!all(numcol[length(numcol)] == "")) {
        numcol <- utils::read.table(file, col.names = 1:initial, 
            fill = TRUE, quote = "", colClasses = "character", 
            nrows = nrows, skip = skip, comment.char = "")
        initial <- initial + 100
    }
    nummax <- max(which(apply(numcol, 2, function(x) all(x == 
        "")) == FALSE)) + 1
    return(nummax)
}
```

Cheers,

**Comment by yukio-takeuchi** [comment]
I have one more suggestion. I remember R has utils::count.fields which, in my sense what we need. It is 10~20 times faster than "revised" get_ncol above  for CompReport.sso with 464,958 lines as
```
> system.time(print(get_ncol("CompReport.sso")))
[1] 106
   user  system elapsed
 18.952   1.754  20.744
> system.time(print(max(count.fields("CompReport.sso"))))
[1] 105
   user  system elapsed
  0.726   0.020   0.748
> fi<-count.fields("CompReport.sso")
> length(fi)
[1] 464958
``` 
The reason why get_ncol return 1 larger than count.fields is the last (most right) column of table by  get_ncol is all blank("") while count.fields counts number of columns of each line.

**Comment by yukio-takeuchi** [comment]
@iantaylor-NOAA 

As for get_ncol, I would like to propose to replace the current get_ncl with following
```
get_ncol <- function(file, skip = 0, nrows = -1) {
  nummax <- count.fields(file, skip = skip, quote ="",
    comment.char = "") + 1
  return(nummax)
}
```
The argument nrow is kept for compatibility. But it is actually not use by the modified one.

Possible advantages are
1. As count.fields returns a vector of length=nrow(file) containing number of columns of each row, this is straight forward.
2. Original get_ncol has a difficulty to set the flexible initial value of ``initial``. One may use read.table without col.names. But without col.names  read.table determine number of columns by reading 25 rows from the top. 25 is hardwired so that if some rows has more columns, it causes an error later.
3. According to the help of count.fields, it was used by older versions of read.table to determine number of columns. I suspect, at that time, count.fields was not fast due to poor computing power and slower disk I/O. Because of that read.table gave up the idea to use it. Given fast computing power and disk I/O, count.field is fast enough.
I personally tested above code so that it was 25 times faster.

**Comment by iantaylor-NOAA** [comment]
Hi @yukio-takeuchi, sorry this took a while for me to get to, but I've just pushed a revision to `get_ncol()` in commit 7c0c382ab5af1146ee60654faf827604ff30e00f to the `get_ncol` branch which uses your code with `count.fields()`. Seems like this makes things both faster and simpler. If it passes the tests, I'll merge it is soon rather than forgetting about it like the last set of changes in that branch which never got merged.
